### DISCUSSION TITLE
How do I get the log index of a LogEntry in the ApplyAsync method of PersistentState

### DISCUSSION THREAD

**Discussion by natilivni** [dotnet/dotNext#33] [Q&A] [answered]
As I understand it, the `ApplyAsync` method of `PersistentState` should be overwritten in any cluster which needs to keep track of its state.  How do I get the log index of the applied message?  Do I have to store it in the message wrapped by LogEntry itself?

**Comment by sakno** [comment, marked as answer]
You don't need to know the index inside of `ApplyAsync`. This method is responsible for interpreting committed entries and applying modifications to the underlying database engine.

**Reply by natilivni** [reply]
There are circumstances where it is useful to know it.  For example, if I am trying to enforce locking in a particular order, it is very useful to know the placement of each message in the audit log and use it to build a current locking state.  In such a scenario, if I create an ordinal number on the client or even in the api, I have no guarantee of the order in which locks were created.  The only source of truth I can have in this case is the audit log index of the message.

**Reply by sakno** [reply]
Only linear interpretation of the committed log entries allow you to restore the state of the underlying database engine. It's bad design to rely on the index. Moreover, snapshot erases all log entries prior to some point in time because the interpreting snapshot has the same effect as interpreting unsquashed log entries.
